python/samples/demos/self_hosted_agent/app/self_hosted_api_chat_completion.py

```diff
@@ -0,0 +1,362 @@
+            if len(chunk.choices) == 0 and chunk.usage is None:
+                continue
+
+            assert isinstance(chunk, ChatCompletionChunk)  # nosec
+            chunk_metadata = self._get_metadata_from_streaming_chat_response(chunk)
+            if chunk.usage is not None:
+                # Usage is contained in the last chunk where the choices are empty
+                # We are duplicating the usage metadata to all the choices in the response
+                yield [
+                    StreamingChatMessageContent(
+                        role=AuthorRole.ASSISTANT,
+                        content="",
+                        choice_index=i,
+                        inner_content=chunk,
+                        ai_model_id=settings.ai_model_id,
+                        metadata=chunk_metadata,
+                        function_invoke_attempt=function_invoke_attempt,
+                    )
+                    for i in range(settings.number_of_responses or 1)
+                ]
+            else:
+                yield [
+                    self._create_streaming_chat_message_content(chunk, choice, chunk_metadata, function_invoke_attempt)
+                    for choice in chunk.choices
+                ]
+
+    @override
+    def _verify_function_choice_settings(self, settings: "PromptExecutionSettings") -> None:
+        if not isinstance(settings, OpenAIChatPromptExecutionSettings):
+            raise ServiceInvalidExecutionSettingsError("The settings must be an OpenAIChatPromptExecutionSettings.")
+        if settings.number_of_responses is not None and settings.number_of_responses > 1:
+            raise ServiceInvalidExecutionSettingsError(
+                "Auto-invocation of tool calls may only be used with a "
+                "OpenAIChatPromptExecutions.number_of_responses of 1."
+            )
+
+    @override
+    def _update_function_choice_settings_callback(
+        self,
+    ) -> Callable[["FunctionCallChoiceConfiguration", "PromptExecutionSettings", FunctionChoiceType], None]:
+        return update_settings_from_function_call_configuration
+
+    @override
+    def _reset_function_choice_settings(self, settings: "PromptExecutionSettings") -> None:
+        if hasattr(settings, "tool_choice"):
+            settings.tool_choice = None
+        if hasattr(settings, "tools"):
+            settings.tools = None
+
+    # endregion
+
+    # region content creation
+    async def _send_completion_request(
+        self,
+        settings: OpenAIChatPromptExecutionSettings,
+    ) -> list[ChatMessageContent]:
+        """Execute the appropriate call to self-hosted agents."""
+        try:
+            settings_dict = settings.prepare_settings_dict()
+            self._handle_structured_output(settings, settings_dict)
+            if settings.tools is None:
+                settings_dict.pop("parallel_tool_calls", None)
+            async with httpx.AsyncClient(timeout=30) as client:
```

Comment:
One can already pass in a custom client and service:

```python
from httpx import AsyncClient

async with AsyncClient(timeout=30) as http_client:
    custom_openai_client = AsyncOpenAI(http_client=http_client)
    custom_openai_service = OpenAIChatCompletion(async_client=custom_openai_client)
    agent = ChatCompletionAgent(
        service=custom_openai_service,
        name="Host",
        instructions="Answer questions about the menu.",
    )
```

By doing so, one can leverage the existing code in the `OpenAIChatCompletionBase`